File: esim-frontend/esim-ide/mainwindow.cc
/* ****************************************************************************
 * OpenESIM (Electrical Engineering Simulator)                                *
 * Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>                      *
 *                                                                            *
 * This project is free software; you can redistribute it and/or              *
 * modify it under the terms of the GNU Lesser General Public License(GPL)    *
 * as published by the Free Software Foundation; either version 2.1           *
 * of the License, or (at your option) any later version.                     *
 *                                                                            *
 * This project is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU          *
 * Lesser General Public License for more details.                            *
 * ****************************************************************************/

#include "mainwindow.h"
#include "schematicview.h"
#include "workspace.h"
#include <QWidget>
#include <QHBoxLayout>
#include <QDesktopWidget>
#include <QRect>
#include <QPoint>
#include <QAction>
#include <QMenu>
#include <QMenuBar>
#include <QToolBar>
#include <QMessageBox>
#include <QFileDialog>
#include <QCursor>
#include <iostream>

namespace esim
{

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
{
  m_workspace = new Workspace(this);

  setWindowTitle("OpenESIM (Electrical Engineering Simulator)");

  createWorkspace();
  createWidgets();
  updateActions();
}

MainWindow::~MainWindow()
{
}

void MainWindow::createWorkspace()
{
  /*
   * File Actions
   */
  m_actionNew = new QAction(QIcon(":/assets/filenew.png"), tr("&New document."), this);
  m_actionNew->setShortcut(QKeySequence::New);
  m_actionNew->setStatusTip(tr("New document."));
  connect(m_actionNew, SIGNAL(triggered(bool)), this, SLOT(slotNew(bool)));

  m_actionOpen = new QAction(QIcon(":/assets/fileopen.png"), tr("&Open a document."), this);
  m_actionOpen->setShortcut(QKeySequence::Open);
  m_actionOpen->setStatusTip(tr("Open a document."));
  connect(m_actionOpen, SIGNAL(triggered(bool)), this, SLOT(slotOpen(bool)));

  m_actionSave = new QAction(QIcon(":/assets/filesave.png"), tr("&Save the current file."), this);
  m_actionSave->setShortcut(QKeySequence::Save);
  m_actionSave->setStatusTip(tr("Save the current file."));
  connect(m_actionSave, SIGNAL(triggered(bool)), this, SLOT(slotSave(bool)));

  m_actionSaveall = new QAction(QIcon(":/assets/filesaveall.png"), tr("&Save All the files."), this);
  m_actionSaveall->setStatusTip(tr("Save All the files."));
  connect(m_actionSaveall, SIGNAL(triggered(bool)), this, SLOT(slotSaveall(bool)));

  m_actionClose = new QAction(QIcon(":/assets/fileclose.png"), tr("&Close the current file."), this);
  m_actionClose->setShortcut(QKeySequence::Close);
  m_actionClose->setStatusTip(tr("Close the current file."));
  connect(m_actionClose, SIGNAL(triggered(bool)), this, SLOT(slotClose(bool)));

  m_actionCloseAll = new QAction(QIcon(":/assets/filecloseall.png"), "&Close All the files.", this);
  m_actionCloseAll->setStatusTip(tr("Close All the files."));
  connect(m_actionCloseAll, SIGNAL(triggered(bool)), this, SLOT(slotCloseAll(bool)));

  m_actionCloseAllButThis = new QAction(tr("&Close All but this file."), this);
  m_actionCloseAllButThis->setStatusTip(tr("Close All the files."));
  connect(m_actionCloseAllButThis, SIGNAL(triggered(bool)), this, SLOT(slotCloseAllButThis(bool)));

  m_actionPrint = new QAction(QIcon(":/assets/print.png"), tr("&Print"), this);
  m_actionPrint->setStatusTip(tr("Print"));
  connect(m_actionPrint, SIGNAL(triggered(bool)), this, SLOT(slotPrint(bool)));

  m_actionPrintSetup = new QAction(QIcon(":/assets/printsetup.png"), tr("&Print Setup"), this);
  m_actionPrintSetup->setStatusTip(tr("Print Setup"));
  connect(m_actionPrintSetup, SIGNAL(triggered(bool)), this, SLOT(slotPrintSetup(bool)));

  m_actionQuit = new QAction(tr("&Exit"), this);
  m_actionQuit->setStatusTip(tr("Exit application"));
  connect(m_actionQuit, SIGNAL(triggered(bool)), this, SLOT(slotQuit(bool)));

  /*
   * Windows Actions
   */
  m_actionGotoPrevTab = new QAction(QIcon(":/assets/gotoprev.png"), tr("&Goto prev Table"), this);
  m_actionGotoPrevTab->setStatusTip(tr("Goto prev Table..."));
  connect(m_actionGotoPrevTab, SIGNAL(triggered(bool)), this, SLOT(slotGotoPrevTab(bool)));
  m_actionGotoNextTab = new QAction(QIcon(":/assets/gotonext.png"), tr("&Goto next Table"), this);
  m_actionGotoNextTab->setStatusTip(tr("Goto next Table..."));
  connect(m_actionGotoNextTab, SIGNAL(triggered(bool)), this, SLOT(slotGotoNextTab(bool)));

  QMenuBar *menubar = menuBar();
  /*
   * Menu - File
   */
  m_menuFile = menubar->addMenu(tr("&File"));
  m_menuFile->addAction(m_actionNew);
  m_menuFile->addSeparator();
  m_menuFile->addAction(m_actionOpen);
  m_menuFile->addAction(m_actionSave);
  m_menuFile->addAction(m_actionSaveall);
  m_menuFile->addSeparator();
  m_menuFile->addAction(m_actionClose);
  m_menuFile->addAction(m_actionCloseAll);
  m_menuFile->addSeparator();
  m_menuFile->addAction(m_actionPrint);
  m_menuFile->addAction(m_actionPrintSetup);
  m_menuFile->addSeparator();
  m_menuFile->addAction(m_actionQuit);

  /*
   * Menu - Edit
   */
  m_menuEdit = menubar->addMenu(tr("&Edit"));

  /*
   * Menu - Windows
   */
  m_menuWindows = menubar->addMenu(tr("&Windows"));
  m_menuWindows->addAction(m_actionClose);
  m_menuWindows->addAction(m_actionCloseAllButThis);
  m_menuWindows->addAction(m_actionCloseAll);
  m_menuWindows->addSeparator();
  m_menuWindows->addAction(m_actionGotoPrevTab);
  m_menuWindows->addAction(m_actionGotoNextTab);

  /*
   * File ToolBar
   */
  m_toolbarFile = addToolBar(tr("File"));
  m_toolbarFile->addAction(m_actionNew);
  m_toolbarFile->addAction(m_actionOpen);
  m_toolbarFile->addAction(m_actionSave);
  m_toolbarFile->addAction(m_actionSaveall);
  m_toolbarFile->addSeparator();
  m_toolbarFile->addAction(m_actionClose);
  m_toolbarFile->addAction(m_actionCloseAll);
  m_toolbarFile->addSeparator();
  m_toolbarFile->addAction(m_actionPrint);
  m_toolbarFile->addAction(m_actionPrintSetup);
  m_toolbarFile->setIconSize( toolIconSize() );
}

void MainWindow::createWidgets()
{
  /*
   * SDI Table
   */
  m_sdiTab = new QTabWidget(this);
  m_sdiTab->setTabsClosable(true);
  connect(m_sdiTab, SIGNAL(currentChanged(int)), this, SLOT(slotCurrentDocumentChanged(int)));
  connect(m_sdiTab, SIGNAL(tabCloseRequested(int)), this, SLOT(slotCloseDocument(int)));

  QWidget *centralWidget = new QWidget(this);
  QHBoxLayout *centralLayout = new QHBoxLayout(centralWidget);
  centralLayout->setContentsMargins(0, 0, 0, 0);
  centralLayout->setSpacing(0);
  centralLayout->addWidget(m_sdiTab);
  setCentralWidget(centralWidget);
}

void MainWindow::updateActions()
{
  bool hasDocuments = (m_workspace->count() > 0);
  bool oneMoreThanDocuments = (m_workspace->count() > 1);
  m_actionSave->setEnabled(hasDocuments);
  m_actionSaveall->setEnabled(hasDocuments);
  m_actionPrint->setEnabled(hasDocuments);

  m_actionClose->setEnabled(hasDocuments);
  m_actionCloseAll->setEnabled(oneMoreThanDocuments);
  m_actionCloseAllButThis->setEnabled(oneMoreThanDocuments);
  m_actionGotoPrevTab->setEnabled(oneMoreThanDocuments);
  m_actionGotoNextTab->setEnabled(oneMoreThanDocuments);
}

QSize MainWindow::sizeHint () const {
  return QSize(1400, 770);
}

void MainWindow::center()
{
  QRect qr(frameGeometry());
  QPoint cp(QDesktopWidget().availableGeometry().center());
  qr.moveCenter(cp);
  move(qr.topLeft());
}

WorkspaceFixed *MainWindow::newDocument(const char *filename)
{
  WorkspaceClass wclass = m_workspace->getClassByFilename(filename);
  if (wclass != WORKSPACE_UNKNOWN)
    {
      WorkspaceFixed *ws = m_workspace->create(wclass, filename);
      m_sdiTab->addTab(ws->view(), QString("%1 - [%2]").arg(filename).arg(m_workspace->count()-1));
      m_sdiTab->setCurrentWidget(ws->view());
      updateActions();
      return ws;
    }
  return 0l;
}

void MainWindow::closeDocument(int tabIndex)
{
  WorkspaceFixed *ws = m_workspace->workspace(tabIndex);
  if (ws->modified())
    {
      int btn = QMessageBox::warning(this,
          tr("Close document"),
          QString(tr("The document '%1' has been modified, would you want to save it?")).arg(ws->filename()),
          QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes);

      if (btn == QMessageBox::Yes)
        { ws->save(); }
    }
  m_workspace->remove(ws);
  m_sdiTab->removeTab(m_sdiTab->indexOf(ws->view()));
  updateActions();
}

void MainWindow::openDocument(const char *filename)
{
  if (WorkspaceFixed *ws = newDocument(filename))
    { ws->open(); }
}

void MainWindow::saveDocument(int tabIndex)
{
  WorkspaceFixed *ws = m_workspace->workspace(tabIndex);
  ws->save();
}

void MainWindow::gotoPrevTab()
{
  int last = m_workspace->count();
  int current = m_sdiTab->currentIndex();
  if (last)
    {
      last -= 1;
      current = (current == 0) ? last : current - 1;
      m_sdiTab->setCurrentIndex(current);
    }
}

void MainWindow::gotoNextTab()
{
  int last = m_workspace->count();
  int current = m_sdiTab->currentIndex();
  if (last)
    {
      last -= 1;
      current = (current == last) ? 0 : current + 1;
      m_sdiTab->setCurrentIndex(current);
    }
}

/**
 * @brief Activate workspace.
 * @param index Index of target workspace
 */
void MainWindow::activateWorkspace(int index)
{
  for(int i=0; i < m_workspace->count(); i++)
    {
      WorkspaceFixed *currentDocument = m_workspace->workspace(i);
      currentDocument->activate(i==index); // Activate the current workspace while disabling others.
    }
  updateActions();
}

/*
 * Slot handlers
 */
void MainWindow::slotCurrentDocumentChanged(int index)
{
  if (index != -1)
    { activateWorkspace(index); }
  updateActions();
}

void MainWindow::slotCloseDocument(int index)
{
  if (index != -1)
    { closeDocument(index); }
}

void MainWindow::slotNew(bool checked)
{
  newDocument("./sch1.sch");
}
void MainWindow::slotOpen(bool checked)
{
  QString filename = QFileDialog::getOpenFileName(this, tr("Open file"));
  if (filename.length())
    { openDocument(filename.toLatin1().data()); }
}
void MainWindow::slotSave(bool checked)
{
  saveDocument(m_sdiTab->currentIndex());
}
void MainWindow::slotSaveall(bool checked)
{
  for (int i=0; i < m_workspace->count(); i++)
    saveDocument(i);
}
void MainWindow::slotClose(bool checked)
{
  closeDocument(m_sdiTab->currentIndex());
}
void MainWindow::slotCloseAll(bool checked)
{
  while (m_workspace->count() > 0)
    { closeDocument(0); }
}
void MainWindow::slotCloseAllButThis(bool checked)
{
  int thisId = m_sdiTab->currentIndex();
  int i = 0;
  while (m_sdiTab->count() > 1)
    {
      if (i < thisId)
        { closeDocument(0); }
      else if (i > thisId)
        { closeDocument(1); }
      i += 1;
    }
}
void MainWindow::slotPrint(bool checked)
{
}
void MainWindow::slotPrintSetup(bool checked)
{
}
void MainWindow::slotQuit(bool checked)
{
}
void MainWindow::slotGotoPrevTab(bool chekced)
{
  gotoPrevTab();
}
void MainWindow::slotGotoNextTab(bool checked)
{
  gotoNextTab();
}

} // namespace esim

File: esim-frontend/test/mainWindow.py
#
#  OpenESIM (Electrical Engineering Simulator)
#  Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>
#
#  This project is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public License(GPL)
#  as published by the Free Software Foundation; either version 2.1
#  of the License, or (at your option) any later version.
#
#  This project is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#

from PyQt5.QtWidgets import QApplication, QAction, QWidget, QHBoxLayout, QMainWindow, QTreeWidget, QTreeWidgetItem, QMessageBox, QDesktopWidget, QFileDialog
from PyQt5.QtWidgets import QTableWidget, QTableWidgetItem, QHeaderView
from PyQt5.QtWidgets import QTabWidget
from PyQt5.QtGui import QIcon, QColor
from PyQt5.QtCore import Qt, QSize
from workspace import DocumentType
from workspaceSchematic import WorkspaceSchematic
from schematic.schematicWidget import SchematicWidget
import sip
import objgraph

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        
        self.setGeometry(300, 300, 300, 220)
        self.setWindowTitle('OpenESIM (Electrical Engineering Simulator)')
        
        #
        # Actions of workspace
        #
        self.actionNew = QAction(QIcon(':/assets/filenew.png'), '&New document.', self)
        self.actionNew.setShortcut('Ctrl+N')
        self.actionNew.setStatusTip('New document.')
        self.actionNew.triggered.connect(self.slotNew)
        
        self.actionOpen = QAction(QIcon(':/assets/fileopen.png'), '&Open a document.', self)
        self.actionOpen.setShortcut('Ctrl+O')
        self.actionOpen.setStatusTip('Open a document.')
        self.actionOpen.triggered.connect(self.slotOpen)
        
        self.actionSave = QAction(QIcon(':/assets/filesave.png'), '&Save the current file.', self)
        self.actionSave.setShortcut('Ctrl+S')
        self.actionSave.setStatusTip('Save the current file.')
        self.actionSave.triggered.connect(self.slotSave)
        
        self.actionSaveall = QAction(QIcon(':/assets/filesaveall.png'), '&Save All the files.', self)
        self.actionSaveall.setShortcut('Ctrl+Shift+S')
        self.actionSaveall.setStatusTip('Save All the files.')
        self.actionSaveall.triggered.connect(self.slotSaveall)
        
        self.actionClose = QAction(QIcon(':/assets/fileclose.png'), '&Close the current file.', self)
        self.actionClose.setShortcut('Ctrl+K')
        self.actionClose.setStatusTip('Close the current file.')
        self.actionClose.triggered.connect(self.slotClose)
        
        self.actionCloseAll = QAction(QIcon(':/assets/filecloseall.png'), '&Close All the files.', self)
        self.actionCloseAll.setShortcut('Ctrl+ShiftK')
        self.actionCloseAll.setStatusTip('Close All the files.')
        self.actionCloseAll.triggered.connect(self.slotCloseAll)
        
        self.actionCloseAllButThis = QAction('&Close All but this file.', self)
        self.actionCloseAllButThis.setShortcut('Ctrl+ShiftK')
        self.actionCloseAllButThis.setStatusTip('Close All the files.')
        self.actionCloseAllButThis.triggered.connect(self.slotCloseAllButThis)
        
        self.actionPrint = QAction(QIcon(':/assets/print.png'), '&Print', self)
        self.actionPrint.setStatusTip('Print')
        self.actionPrint.triggered.connect(self.slotPrint)
        
        self.actionPrintSetup = QAction(QIcon(':/assets/printsetup.png'), '&Print Setup', self)
        self.actionPrintSetup.setStatusTip('Print Setup')
        self.actionPrintSetup.triggered.connect(self.slotPrintSetup)
        
        self.actionQuit = QAction('&Exit', self)
        self.actionQuit.setShortcut('Ctrl+Q')
        self.actionQuit.setStatusTip('Exit application')
        self.actionQuit.triggered.connect(self.slotQuit)

        self.actionGotoPrevTab = QAction(QIcon(':/assets/gotoprev.png'), '&Goto prev Table', self)
        self.actionGotoPrevTab.setStatusTip('Goto prev Table...')
        self.actionGotoPrevTab.triggered.connect(self.slotGotoPrevTab)
        self.actionGotoNextTab = QAction(QIcon(':/assets/gotonext.png'), '&Goto next Table', self)
        self.actionGotoNextTab.setStatusTip('Goto next Table...')
        self.actionGotoNextTab.triggered.connect(self.slotGotoNextTab)
 
        self.statusBar()
 
        menubar = self.menuBar()
        #
        # Menu - File
        #
        self.menuFile = menubar.addMenu('&File')
        self.menuFile.addAction(self.actionNew)
        self.menuFile.addSeparator()
        self.menuFile.addAction(self.actionOpen)
        self.menuFile.addAction(self.actionSave)
        self.menuFile.addAction(self.actionSaveall)
        self.menuFile.addSeparator()
        self.menuFile.addAction(self.actionClose)
        self.menuFile.addAction(self.actionCloseAll)
        self.menuFile.addSeparator()
        self.menuFile.addAction(self.actionPrint)
        self.menuFile.addAction(self.actionPrintSetup)
        self.menuFile.addSeparator()
        self.menuFile.addAction(self.actionQuit)
        
        #
        # Menu - Edit
        #
        self.menuEdit = menubar.addMenu('&Edit')
        
        #
        # Menu - Windows
        #
        self.menuWindows = menubar.addMenu('&Windows')
        self.menuWindows.addAction(self.actionClose)
        self.menuWindows.addAction(self.actionCloseAllButThis)
        self.menuWindows.addAction(self.actionCloseAll)
        self.menuWindows.addSeparator()
        self.menuWindows.addAction(self.actionGotoPrevTab)
        self.menuWindows.addAction(self.actionGotoNextTab)
        
        #
        # File ToolBar
        #
        self.toolbarFile = self.addToolBar('File')
        self.toolbarFile.addAction(self.actionNew)
        self.toolbarFile.addAction(self.actionOpen)
        self.toolbarFile.addAction(self.actionSave)
        self.toolbarFile.addAction(self.actionSaveall)
        self.toolbarFile.addSeparator()
        self.toolbarFile.addAction(self.actionClose)
        self.toolbarFile.addAction(self.actionCloseAll)
        self.toolbarFile.addSeparator()
        self.toolbarFile.addAction(self.actionPrint)
        self.toolbarFile.addAction(self.actionPrintSetup)
        self.toolbarFile.setIconSize( QSize(20, 29) )
        
        #
        # Edit ToolBar
        #
        self.toolbarEdit = self.addToolBar('Edit')
        
        self.setGeometry(300, 300, 1600, 800)
        self.center()
        
        #
        # TreeView for database entries
        #
        self.layoutTreeWidget = QTreeWidget()
        self.layoutTreeWidget.setColumnCount(2)
        self.layoutTreeWidget.setHeaderLabels(['Database','Filename'])
        self.layoutTreeWidget.header().setSectionResizeMode(QHeaderView.ResizeToContents);
        
        self.layoutTreeGroups = QTreeWidgetItem(self.layoutTreeWidget)
        self.layoutTreeGroups.setText(0, 'Groups')
        self.layoutTreeGroups.setExpanded(True)
        self.layoutTreeWidget.addTopLevelItem(self.layoutTreeGroups)
        
        #
        # SDI Table
        #
        self.tabWidget = QTabWidget()
        self.tabWidget.setTabsClosable(True)
        self.tabWidget.currentChanged.connect(self.slotCurrentDocumentChanged)
        self.tabWidget.tabCloseRequested.connect(self.closeDocument)

        self.centralWidget = QWidget(self)
        self.centralLayout = QHBoxLayout(self.centralWidget)
        self.centralLayout.setContentsMargins(0, 0, 0, 0)
        self.centralLayout.setSpacing(0)
        self.centralLayout.addWidget(self.layoutTreeWidget)
        self.centralLayout.addWidget(self.tabWidget)
        self.centralLayout.setStretchFactor(self.layoutTreeWidget, 1);
        self.centralLayout.setStretchFactor(self.tabWidget, 5);
        self.setCentralWidget(self.centralWidget)
        
        #
        # Prepare Workspaces
        #
        self.workspaces = {}
        self.workspaces[int(DocumentType.DOC_SCHEMATIC)] = WorkspaceSchematic(self, False)
        self.workspaces[int(DocumentType.DOC_SCHEMATIC_SYMBOL)] = WorkspaceSchematic(self, True)
        self.documents = []
        self.currentDocument = None
        self.currentWorkspace = None
        self.__activateWorkspace(None)
        
    def updateActions(self):
        hasDocuments = (self.tabWidget.count() > 0)
        oneMoreThanDocuments = self.tabWidget.count() > 1
        self.actionSave.setEnabled(hasDocuments)
        self.actionSaveall.setEnabled(hasDocuments)
        self.actionPrint.setEnabled(hasDocuments)
        
        self.actionClose.setEnabled(hasDocuments)
        self.actionCloseAll.setEnabled(oneMoreThanDocuments)
        self.actionCloseAllButThis.setEnabled(oneMoreThanDocuments)
        self.actionGotoPrevTab.setEnabled(oneMoreThanDocuments)           
        self.actionGotoNextTab.setEnabled(oneMoreThanDocuments)
        
    def center(self):
        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())
    
    def workspace(self, type):
        return self.workspaces[int(type)]
    
    def newDocument(self, filename):
        staticWS = self.workspace( DocumentType.getFromFilename(filename) )
        
        document = staticWS.createDocument(self, filename)
        self.documents.append(document)
        
        self.tabWidget.addTab(document.view, '%s - [%d]' % (filename, len(self.documents)-1) )
        self.tabWidget.setCurrentWidget(document.view)
        
        return document
        
    def closeDocument(self, tabIndex):
        document = self.documents[tabIndex]
        if document.modified():
            btn = QMessageBox.warning(self,
                                    'Close document', "The document '%s' has been modified, would you want to save it?" % document.filename,
                                    QMessageBox.Yes | QMessageBox.No, QMessageBox.Yes)
            if btn == QMessageBox.Yes:
                document.save()

        document.close()
        self.tabWidget.removeTab(self.tabWidget.indexOf(document.view))
        view = document.view
        self.documents.remove(document)
        sip.delete(view) #!FIXME ugly delete C++ pointer here, but memory leaked without sake?

    def closeAllDocuments(self):
        while self.tabWidget.count() > 0:
            self.closeDocument(0)
        
    def openDocument(self, filename):
        document = self.newDocument(filename)
        result = document.open()

    def saveDocument(self, tabIndex):
        document = self.newDocument(filename)
        result = document.save()

    def gotoPrevTab(self):
        last = self.tabWidget.count()
        current = self.tabWidget.currentIndex()
        if last:
            last -= 1
            current = last if current == 0 else current - 1
            self.tabWidget.setCurrentIndex(current)
            
    def gotoNextTab(self):
        last = self.tabWidget.count()
        current = self.tabWidget.currentIndex()
        if last:
            last -= 1
            current = 0 if current == last else current + 1
            self.tabWidget.setCurrentIndex(current)

    """
    @brief Activate workspace.
    @param currentWorkspace Point to the target workspace, None if nothing.
    """
    def __activateWorkspace(self, currentWorkspace, currentDocument=None):
        if currentWorkspace is not None:
            for type, staticWS in self.workspaces.items():
                if currentWorkspace is staticWS:
                    staticWS.switch(True, currentDocument)
                else:
                    staticWS.switch(False, None)
        else:
            for type, staticWS in self.workspaces.items():
                staticWS.switch(False, None)
        
        self.currentDocument = currentDocument
        self.currentWorkspace = currentWorkspace
        
        self.updateActions()
    
    #
    # Slot Processing
    #
    def slotNew(self, event):
        for i in range(1):
            self.newDocument("./sch1.sch")
    def slotOpen(self, event):
        filename, types = QFileDialog.getOpenFileName(self,'Open file')
        if len(filename):
            self.openDocument(filename)
    def slotSave(self, event):
        self.saveDocument(self.tabWidget.currentIndex())
    def slotSaveall(self, event):
        for i in range(self.tabWidget.count()):
            self.saveDocument(i)
    def slotClose(self, event):
        self.closeDocument(self.tabWidget.currentIndex())
        #objgraph.show_backrefs(objgraph.by_type('DocumentSchematic')[0], max_depth = 10, filename = 'obj.dot')
    def slotCloseAll(self, event):
        self.closeAllDocuments()
    def slotCloseAllButThis(self, event):
        thisId = self.tabWidget.currentIndex()
        i = 0
        while self.tabWidget.count() > 1:
            if i < thisId:
                self.closeDocument(0)
            elif i > thisId:
                self.closeDocument(1)
            i += 1

    def slotPrint(self, event):
        pass
    def slotPrintSetup(self, event):
        pass
         
    def slotQuit(self, event):
        self.close()

    def slotGotoPrevTab(self, event):
        self.gotoPrevTab()
    def slotGotoNextTab(self, event):
        self.gotoNextTab()
    
    def treeItemDoubleClick(self, item, index):
        pass
    
    def slotCurrentDocumentChanged(self, index): # Handle switching among documents
        if self.currentDocument is not None:
            self.currentDocument.activate(False) # restore the status of previous document
        if index != -1:
            currentDocument = self.documents[index]
            currentDocument.activate(True) # Activate the document and workspace
            self.__activateWorkspace(currentDocument.workspace, currentDocument)
        else:
            self.__activateWorkspace(None)

    def closeEvent(self, event):
        self.closeAllDocuments()
